clients/client-rds-data/RdsDataServiceClient.ts

```diff
@@ -0,0 +1,206 @@
+   */
+  sha256?: HashConstructor;
+
+  /**
+   * Default credentials provider; Not available in browser runtime
+   */
+  credentialDefaultProvider?: (input: any) => Provider<Credentials>;
+
+  /**
+   * Provider function that return promise of a region string
+   */
+  regionDefaultProvider?: (input: any) => Provider<string>;
+
+  /**
+   * The function that will be used to convert strings into HTTP endpoints
+   */
+  urlParser?: UrlParser;
+
+  /**
+   * A function that can calculate the length of a request body.
+   */
+  bodyLengthChecker?: (body: any) => number | undefined;
+
+  /**
+   * A function that converts a stream into an array of bytes.
+   */
+  streamCollector?: StreamCollector;
+
+  /**
+   * The function that will be used to convert a base64-encoded string to a byte array
+   */
+  base64Decoder?: Decoder;
+
+  /**
+   * The function that will be used to convert binary data to a base64-encoded string
+   */
+  base64Encoder?: Encoder;
+
+  /**
+   * The function that will be used to convert a UTF8-encoded string to a byte array
+   */
+  utf8Decoder?: Decoder;
+
+  /**
+   * The function that will be used to convert binary data to a UTF-8 encoded string
+   */
+  utf8Encoder?: Encoder;
+
+  /**
+   * The function that will be used to populate default value in 'User-Agent' header
+   */
+  defaultUserAgent?: string;
+
+  /**
+   * The function that will be used to populate serializing protocol
+   */
+  protocolDefaultProvider?: (
+    handler: HttpHandler
+  ) => Protocol<HttpRequest, HttpResponse>;
+
+  /**
+   * The service name with which to sign requests.
+   */
+  signingName?: string;
+
+  /**
+   * The service name with which to construct endpoints.
+   */
+  service?: string;
+}
+
+export type RdsDataServiceConfig = RDSDataRuntimeDependencies &
+  AwsAuthConfigInput &
+  RegionConfigInput &
+  RetryConfigInput &
+  EndpointsConfigInput &
+  ClientProtocolConfigInput &
+  UserAgentConfigInput;
+
+export type RdsDataServiceResolvedConfig = SmithyResolvedConfiguration<
+  __HttpOptions
+> &
+  Required<RDSDataRuntimeDependencies> &
+  AwsAuthConfigResolved &
+  RegionConfigResolved &
+  RetryConfigResolved &
+  EndpointsConfigResolved &
+  ClientProtocolConfigResolved &
+  UserAgentConfigResolved;
+
+export class RdsDataService extends SmithyClient<
+  __HttpOptions,
+  ServiceInputTypes,
+  ServiceOutputTypes
+> {
+  readonly config: RdsDataServiceResolvedConfig;
+
+  constructor(configuration: RdsDataServiceConfig) {
+    const _config_0 = resolveClientProtocolConfig({
+      ...RDSRuntimeConfiguration,
+      ...configuration
+    });
+    let _config_1 = resolveRegionConfig(_config_0);
+    let _config_2 = resolveAwsAuthConfig(_config_1);
+    let _config_3 = resolveEndpointsConfig(_config_2);
+    let _config_4 = resolveRetryConfig(_config_3);
+    let _config_5 = resolveUserAgentConfig(_config_4);
+    super(_config_5);
+    this.config = _config_5;
```

Comment:
nit: The below suggestion avoids creation of temporary variables. Can we use it instead if it's not difficult to CodeGen?
```javascript
this.config = resolveUserAgentConfig(
  resolveRetryConfig(
    resolveEndpointsConfig(
      resolveAwsAuthConfig(
        resolveRegionConfig(
          resolveClientProtocolConfig({
            ...RDSRuntimeConfiguration,
            ...configuration
          })
        )
      )
    )
  )
);
super(this.config);

```

Reply:
A simpler way would be to use generic compose function for this ([docs](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#higher-order-type-inference-from-generic-functions))

Needs testing:
```typescript
function compose<A, B, C>(f: (arg: A) => B, g: (arg: B) => C): (arg: A) => C {
    return x => g(f(x));
}

this.config = compose(
  resolveUserAgentConfig,
  resolveRetryConfig,
  resolveEndpointsConfig,
  resolveAwsAuthConfig,
  resolveRegionConfig,
  resolveClientProtocolConfig
)({
  ...RDSRuntimeConfiguration,
  ...configuration
});

```

Reply:
I think this compose interface needs [Variadic Kinds](https://github.com/Microsoft/TypeScript/issues/5453) to yet implemented in TS. Currently there's no way to spreading the generic in TS IMO. For example:

```typescript
function compose<...T>//and other things
```